libraries/AC_PID/AC_PID.h

```diff
@@ -9,102 +9,128 @@
-    //  this should be called before any other calls to get_p, get_i or get_d
-    void        set_input_filter_d(float input);
+    void set_dt(float dt);
+
+    //  update_all - set target and measured inputs to PID controller and calculate outputs
+    //  target and error are filtered
+    //  the derivative is then calculated and filtered
+    //  the integral is then updated based on the setting of the limit flag
+    float update_all(float target, float measurement, bool limit = false);
+
+    //  update_error - set error input to PID controller and calculate outputs
+    //  target is set to zero and error is set and filtered
+    //  the derivative then is calculated and filtered
+    //  the integral is then updated based on the setting of the limit flag
+    //  Target and Measured must be set manually for logging purposes.
+    // todo: remove function when it is no longer used.
+    float update_error(float error, bool limit = false);
+
+    //  update_i - update the integral
+    //  if the limit flag is set the integral is only allowed to shrink
+    void update_i(bool limit);
 
     // get_pid - get results from pid controller
-    float       get_pid();
-    float       get_pi();
-    float       get_p();
-    float       get_i();
-    float       get_d();
-    float       get_ff(float requested_rate);
+    float get_pid() const;
+    float get_pi() const;
+    float get_p() const;
+    float get_i() const;
+    float get_d() const;
+    float get_ff();
+
+    // todo: remove function when it is no longer used.
+    float get_ff(float target);
 
     // reset_I - reset the integrator
-    void        reset_I();
+    void reset_I();
 
     // reset_filter - input filter will be reset to the next value provided to set_input()
-    void        reset_filter() { _flags._reset_filter = true; }
+    void reset_filter() {
+        _flags._reset_filter = true;
+    }
 
     // load gain from eeprom
-    void        load_gains();
+    void load_gains();
 
     // save gain to eeprom
-    void        save_gains();
+    void save_gains();
 
     /// operator function call for easy initialisation
-    void operator() (float p, float i, float d, float imaxval, float input_filt_hz, float dt, float ffval = 0);
+    void operator()(float p_val, float i_val, float d_val, float ff_val, float imax_val, float input_filt_T_hz, float input_filt_E_hz, float input_filt_D_hz, float dt);
 
     // get accessors
-    AP_Float   &kP() { return _kp; }
-    AP_Float   &kI() { return _ki; }
-    AP_Float   &kD() { return _kd; }
-    AP_Float   &filt_hz() { return _filt_hz; }
-    float       imax() const { return _imax.get(); }
-    float       get_filt_alpha() const;
-    float       ff() const { return _ff.get(); }
+    AP_Float &kP() { return _kp; }
+    AP_Float &kI() { return _ki; }
+    AP_Float &kD() { return _kd; }
+    AP_Float &ff() { return _kff;}
+    AP_Float &filt_T_hz() { return _filt_T_hz; }
```

Comment:
no a blocker for this PR but we don't actually use these new accessors so they could be removed.  It might be good to be sure that we don't need to set the defaults in Copter's allocate_motors() function which does seem to set some defaults for kP, kI and dD.